pkg/chunk/local/boltdb_index_client.go

```diff
@@ -117,41 +119,45 @@ func (b *boltIndexClient) Stop() {
 	b.wait.Wait()
 }
 
-func (b *boltIndexClient) NewWriteBatch() chunk.WriteBatch {
+func (b *BoltIndexClient) NewWriteBatch() chunk.WriteBatch {
 	return &boltWriteBatch{
 		tables: map[string]map[string][]byte{},
 	}
 }
 
-func (b *boltIndexClient) getDB(name string) (*bbolt.DB, error) {
+func (b *BoltIndexClient) GetDB(name string, operation int) (*bbolt.DB, error) {
 	b.dbsMtx.RLock()
 	db, ok := b.dbs[name]
 	b.dbsMtx.RUnlock()
 	if ok {
 		return db, nil
 	}
 
-	b.dbsMtx.Lock()
-	defer b.dbsMtx.Unlock()
-	db, ok = b.dbs[name]
-	if ok {
-		return db, nil
-	}
+	if _, err := os.Stat(path.Join(b.cfg.Directory, name)); err == nil || operation == DBOperationWrite {
+		b.dbsMtx.Lock()
+		defer b.dbsMtx.Unlock()
+		db, ok = b.dbs[name]
+		if ok {
+			return db, nil
+		}
 
-	// Open the database.
-	// Set Timeout to avoid obtaining file lock wait indefinitely.
-	db, err := bbolt.Open(path.Join(b.cfg.Directory, name), 0666, &bbolt.Options{Timeout: 5 * time.Second})
-	if err != nil {
-		return nil, err
+		// Open the database.
+		// Set Timeout to avoid obtaining file lock wait indefinitely.
+		db, err := OpenBoltdbFile(path.Join(b.cfg.Directory, name))
+		if err != nil {
+			return nil, err
+		}
+
+		b.dbs[name] = db
+		return db, nil
 	}
 
-	b.dbs[name] = db
-	return db, nil
+	return nil, nil
```

Comment:
We should never return `nil, nil`. Either a DB or an error (an error could be something like "unsupported operation by the configured mode"). The rest of the code doesn't currently handles the case both the DB and thr error are `nil`, and I belive it's correct they don't handle it.

Reply:
for read operation it does check whether db is there or not for querying https://github.com/cortexproject/cortex/pull/2049/files/6cab326d3ae40257ce2a40578b9e2cbcc0afce8c#diff-f0b676dad4198715e462b2faa06cfaa8R195

I don't feel right throwing an error because db does not exist and we don't want to create it for a read operation because it would anyways return no results.

Reply:
> I don't feel right throwing an error

My perspective: if a BoltDB client instance is not configured for read operations, no read operation should occur on it. If it occurs, it's an error, and we should return an error. This also doesn't force us to change all other functions logic to handle the case `GetDB()` returns `nil, nil`.